pkg/server/proto/server.proto

```diff
@@ -4344,6 +4375,124 @@ message EntrypointExecResponse {
   }
 }
 
+/********************************************************************
+* Pipelines
+********************************************************************/
+
+message Pipeline {
+  // Unique ID for this pipeline.
+  string id = 1;
+
+  // Human-friendly name for the pipeline. This is not unique! Users should
+  // use the ID and other metadata to differentiate pipelines. In general,
+  // we recommend avoiding reusing names.
+  string name = 2;
+
+  // The owner is the thing that manages this pipeline.
+  //
+  // NOTE(mitchellh): We currently only support project-synced pipelines.
+  // In the future, we may want to consider pipelines not attached to
+  // projects (just created via the API/UI/CLI on their own). This "oneof"
+  // is here to represent that. If we ever decide NOT to do that, we can just
+  // remove the "oneof" which is backwards compatible.
+  oneof owner {
+    // This pipeline is created from a waypoint.hcl file that is synced via VCS.
+    Ref.Project project = 3;
+  }
+
+  // The steps that make up this pipeline. This is a set and the order
+  // has no meaning; it does not imply any ordering that the steps will
+  // be executed. The execution order is purely defined by the "depends_on"
+  // fields on the steps.
+  //
+  // There must be only one "root" step. A root step is a step with no
+  // dependencies. It is an error for there to be more than one root step.
+  //
+  // For API users, the GetPipelineResponse has a number of fields that
+  // help make this much easier to consume, such as a dedicated field for
+  // the root step, mermaid-formatted graph output, and more.
+  //
+  // These requirements are usually hidden from users through nicer user
+  // experiences for editing: waypoint.hcl syntax (which implicitly creates
+  // ordering and root steps), UIs, etc. All these requirements are noted for
+  // _internal_ users who are modifying pipelines via the API. For those
+  // internal users, they must manage the tedium of adhering to these requirements.
+  map<string, Step> steps = 4;
+
+  // A single step within the pipeline.
+  message Step {
+    // Name of the step. This is unique within a pipeline. In the "steps"
+    // map, this must match the map key.
+    string name = 1;
+
+    // The steps that must come before this step. If this is empty,
+    // this is considered a "root" step. Only one step may be root.
+    // See "steps" on the Pipeline message for details.
+    repeated string depends_on = 2;
```

Comment:
Thinking about this more, do you think we'll ever want more detail from the dependent stages? I'm imagining in the future we might want to depend on a stage even if it fails, or maybe explicitly map it's outputs to our inputs.

Something like this would be more flexible:
```
repeated DependentStep depends_on = 2;

message DependentStep {
  string name = 4;
  bool ignore_failure = 5;
  // ... other future ideas, like explicit output->input mapping
}

```

Reply:
We have some precedent for this already on jobs. The job dependencies are just specified in two separate `repeated string` fields. I do like your idea of making them structured, however. Let me think about this and I might make that change.